xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,288 @@
+          if (clusterWatcher == null || clusterWatcher.edsBalancer == null) {
+            // Go into TRANSIENT_FAILURE if we have not yet received any cluster resource.
+            // Otherwise, we keep running with the data we had previously.
+            helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+          }
+        }
+
+        @Override
+        public boolean canHandleEmptyAddressListFromNameResolution() {
+          return true;
+        }
+
+        @Override
+        public void shutdown() {
+          if (clusterWatcher != null) {
+            if (clusterWatcher.edsBalancer != null) {
+              clusterWatcher.edsBalancer.shutdown();
+            }
+            xdsClient.cancelClusterDataWatch(cdsConfig.name, clusterWatcher);
+          }
+        }
+
+        @Override
+        public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+          if (clusterWatcher == null) {
+            clusterWatcher = new ClusterWatcherImpl(helper, resolvedAddresses);
+            xdsClient.watchClusterData(cdsConfig.name, clusterWatcher);
+            if (oldCdsConfig != null) {
+              xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+            }
+            CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+          }
+          // Else handleResolvedAddresses() already called, so no-op here because the config is
+          // fixed in this balancer.
+        }
+      };
+    }
+  }
+
+  private final class ClusterWatcherImpl implements ClusterWatcher {
+
+    final Helper helper;
+    final ResolvedAddresses resolvedAddresses;
+
+    // EDS balancer for the cluster.
+    // Becomes non-null once handleResolvedAddresses() successfully.
+    // Assigned at most once.
+    @Nullable
+    LoadBalancer edsBalancer;
+
+    // LoadStatsStore for the cluster.
+    // Becomes non-null once handleResolvedAddresses() successfully.
+    // Assigned at most once.
+    @Nullable
+    LoadStatsStore loadStatsStore;
+
+    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
+      this.helper = helper;
+      this.resolvedAddresses = resolvedAddresses;
+    }
+
+    @Override
+    public void onClusterChanged(ClusterUpdate newUpdate) {
+      channelLogger.log(ChannelLogLevel.DEBUG, "Received a CDS update: '%s'",  newUpdate);
+      checkArgument(
+          newUpdate.getLbPolicy().equals("round_robin"),
+          "The load balancing policy in ClusterUpdate '%s' is not supported", newUpdate);
+
+      final XdsConfig edsConfig = new XdsConfig(
+          /* balancerName = */ null,
+          new LbConfig(newUpdate.getLbPolicy(), ImmutableMap.<String, Object>of()),
+          /* fallbackPolicy = */ null,
+          /* edsServiceName = */ newUpdate.getEdsServiceName(),
+          /* lrsServerName = */ newUpdate.getLrsServerName());
+
+      if (loadStatsStore == null) {
+        loadStatsStore = new LoadStatsStoreImpl();
```

Comment:
I am still doubt for the storage unit of a `LoadStatsStore` instance, whether it is for a CDS cluster or an EDS service. So when shall we create a `LoadStatsStore` is still unclear. Need to figure that out first.

Reply:
I think for now, we can only create a `LoadStatsStore` in EDS balancer locally and support only the basic case. I will not create it here. Things may change in the future and it's totally uncertain now.

Reply:
Yes, by no means it should be here. The safest and most correct (if I understand the API correctly) place is at the balancer for EDS service.